ci/cron/src/Main.hs

```diff
@@ -0,0 +1,254 @@
+import qualified Data.Foldable as Foldable
+import qualified Data.HashMap.Strict as H
+import qualified Data.List as List
+import qualified Data.List.Utils as List
+import qualified Data.List.Split as Split
+import qualified Data.Set as Set
+import qualified Data.Text as Text
+import qualified Network.HTTP.Client as HTTP
+import qualified Network.HTTP.Client.TLS as TLS
+import qualified Network.HTTP.Types.Status
+import qualified System.Environment as Env
+import qualified System.Exit as Exit
+import qualified System.IO.Temp as Temp
+import qualified System.Process as System
+
+(|>) :: a -> (a -> b) -> b
+(|>) x f = f x
+
+shell_exit_code :: String -> IO (Exit.ExitCode, String, String)
+shell_exit_code cmd = do
+    System.readProcessWithExitCode "bash" ["-c", cmd] ""
+
+die :: String -> Exit.ExitCode -> String -> String -> IO a
+die cmd (Exit.ExitFailure exit) out err =
+    Exit.die $ unlines ["Subprocess:",
+                         cmd,
+                        "failed with exit code " <> show exit <> "; output:",
+                        "---",
+                        out,
+                        "---",
+                        "err:",
+                        "---",
+                        err,
+                        "---"]
+die _ _ _ _ = Exit.die "Type system too weak."
+
+
+shell :: String -> IO String
+shell cmd = do
+    (exit, out, err) <- shell_exit_code cmd
+    if exit == Exit.ExitSuccess
+    then return out
+    else do
+        die cmd exit out err
+
+shell_ :: String -> IO ()
+shell_ cmd = do
+    _ <- shell cmd
+    return ()
+
+robustly_download_nix_packages :: IO ()
+robustly_download_nix_packages = do
+    h (10 :: Integer)
+    where
+        cmd = "nix-build nix -A tools -A cached"
+        h n = do
+            (exit, out, err) <- shell_exit_code cmd
+            case (exit, n, "unexpected end-of-file" `List.isInfixOf` err) of
+              (Exit.ExitSuccess, _, _) -> return ()
+              (_, 0, _) -> die cmd exit out err
+              (_, _, True) -> h (n - 1)
+              (_, _, False) -> die cmd exit out err
+
+http_get :: JSON.FromJSON a => String -> IO a
+http_get url = do
+    manager <- HTTP.newManager TLS.tlsManagerSettings
+    request' <- HTTP.parseRequest url
+    -- Be polite
+    let request = request' { HTTP.requestHeaders = [(Insensitive.mk (BS.fromString "User-Agent"), BS.fromString "DAML cron (team-daml-language@digitalasset.com)")] }
+    response <- HTTP.httpLbs request manager
+    let Network.HTTP.Types.Status.Status { Network.HTTP.Types.Status.statusCode = status } = HTTP.responseStatus response
+    let body = JSON.decode $ HTTP.responseBody response
+    case (status, body) of
+      (200, Just body) -> return body
+      _ -> Exit.die $ unlines ["GET \"" <> url <> "\" returned status code " <> show status <> ".",
+                               show $ HTTP.responseBody response]
+
+http_post :: String -> [(String, String)] -> LBS.ByteString -> IO LBS.ByteString
```

Comment:
It seems a bit like you are reinventing a high-level HTTP API. I think it’s small enough that this is fine here but if it gets more complex, it might be worth looking into a library. I’ve liked https://hackage.haskell.org/package/req-2.1.0/docs/Network-HTTP-Req.html in the past.

Reply:
Agreed.